### ISSUE TITLE
variables and implicit arguments

### ISSUE THREAD

**Issue by fpvandoorn** [leanprover/lean#224]
I've started rewriting category.lean using variables and the !-notation. Both features make the library a lot nicer!

https://github.com/fpvandoorn/lean/blob/master/library/algebra/category.lean#L209

Two strange things I noticed this afternoon:

1) In the functor namespace I declare the variables

```
  variables {obC obD obE : Type} {C : category obC} {D : category obD} {E : category obE}
```

However, if I use these variables in `respect_id`, i.e. change line 209-210 to

```
  theorem respect_id (F : C ⇒ D) : Π (a : obC), F (ID a) = id :=
  rec (λ obF homF Hid Hcomp, Hid) F
```

then the lines below it won't be accepted:

```
  variable G : D ⇒ E
  check respect_id G
```

However, the following will be accepted

```
  variable F : C ⇒ D
  check respect_id F
```

It looks like `C` and `D` are still treated as a parameter in some sense.

2) I defined in L215

```
 theorem respect_comp {obC obD : Type} {C : category obC} {D : category obD} (F : C ⇒ D) : Π {{a b c : obC}} (g : hom b c) (f : hom a b),
      F (g ∘ f) = F g ∘ F f 
```

However, if I use it in the definition of `compose` below it, I have to give explicit arguments for `a b c`: 

```
respect_comp F _ _ _ g f
```

**Comment by leodemoura** [comment]
I found out what is going on. The universes associated with `variables {obC obD obE : Type}` are not being abstracted. We have three universes (`l1`, `l2` and `l3`), one for each `ob*`.
In the section, `respect_id` is `respect_id.{l1 l2 l3}`. The variables have been abstracted, but the universes are still there. I will fix that tomorrow.

**Comment by leodemoura** [comment]
fixed by e71d4548def63ed991ed3ae27cdef6720d0035a0 and bf01cfeec566f9b90a712c2daa278cbc7f67227b
